services/scripts.py

```diff
@@ -10,58 +11,120 @@ def __init__(self, row: pd.Series) -> None:
         self.track = row.Track
         self.album = row.Album
         self.artist = row.Artist
-        
+
+        self.musicality = row[MUSICALITY_ROWS_norm].values
+
     def __repr__(self):
-        return f'|{self.artist}: {self.track}|\n'
+        return f"|{self.artist}: {self.track}|\n"
+
+    def eff_distance(self, other: "Song") -> float:
+        """Calculate the effective distance between two songs based on their musicality.
+
+        This method computes the Euclidean distance between the musicality attributes of 
+        the current song and another song. The distance is calculated as the square root 
+        of the sum of the squared differences of their musicality values.
+
+        Args:
+            other: Another instance of the Song class to compare against.
+
+        Returns:
+            The effective distance as a float value.
+        """
+
+        return (
+            sum(
+                (
+                    (other.musicality[i] - self.musicality[i]) ** 2
+                    for i in range(len(self.musicality))
+                )
+            )
+            ** 0.5
+        )
 
 
 class DataBase:
     def __init__(self, filename: str) -> None:
         self._storage = defaultdict(lambda: [])
-        df = pd.read_csv(filename, index_col=0)
-        
-        for _, row in df.iterrows():
-            self.add_song(
-                Song(row)       
+        self._filepath: str = filename
+
+        self.load()
+
+    @property
+    def songs(self) -> list[Song]:
+        return sum(self._storage.values(), start=[])
+
+    @staticmethod
+    def _normalize_musicality_rows(df: pd.DataFrame) -> None:
+        for row in MUSICALITY_ROWS:
+            df.loc[:, f"{row}_norm"] = (df[row] - df[row].min()) / (
+                df[row].max() - df[row].min()
             )
-        
+
+    def load(self) -> None:
+        df = pd.read_csv(self._filepath, index_col=0)
+        df.dropna(subset=MUSICALITY_ROWS, inplace=True)
+        self._normalize_musicality_rows(df)
+        for _, row in df.iterrows():
+            self.add_song(Song(row))
 
     def add_song(self, song: Song) -> None:
         self._storage[song.artist].append(song)
 
-    def get_song(self, song_id: int) -> Song | None:
-        return self._storage.get(song_id)
-
     def search(self, request: str) -> list[Song]:
-        _result = []
-        for song in sum(self._storage.values(), start=[]):
-            if request in song.track:
-                _result.append(song)
-        return _result
-    
-        
-            
-            
-            
-            
-        
+        return [
+            song
+            for song in sum(self._storage.values(), start=[])
+            if request in song.track
+        ]
+
+    def similar_songs(self, song: Song, count: int = 5) -> list[Song]:
```

Comment:
there's some problem with the variable name. we have `song` at the bottom and in the function parameters, so it's not good

Reply:
I dont realy get what you mean, can you please explain it in other words?

Reply:
> I dont realy get what you mean, can you please explain it in other words?

in this case we have `song` as the function argument, but bottom we also have a `song` as a variable 
```py
def similar_songs(self, song: Song, count: int = 5) -> list[Song]:  # here
#                       ^^^^^^^^^^
```
```py
if __name__ == "__main__":
    db = DataBase("./data/Spotify_Youtube.csv")
    if song := db._storage["Amadeus"][0]:  # <--- and here
    #  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        print(db.similar_songs(song))
```
there may be confusion or mistakes in the use of variables, so that some interpreters will tag it as pep8 error. It's no big deal, but I try not to do that